src/main/java/os/failsafe/executor/PersistentQueue.java

```diff
@@ -30,37 +34,36 @@ String add(Connection connection, Task task) {
         return taskRepository.add(connection, task).getId();
     }
 
-    Task peekAndLock(Set<String> processableTasks, int limit) {
-        List<Task> nextTasksToLock = findNextForExecution(processableTasks, limit);
+    List<Task> peekAndLock(Set<String> processableTasks, int limit) {
+        return database.connect(con -> {
+            List<Task> nextTasksToLock = findNextForExecution(con, processableTasks, limit);
 
-        if (nextTasksToLock.isEmpty()) {
-            return null;
-        }
-
-        if (Thread.currentThread().isInterrupted()) {
-            return null;
-        }
-
-        do {
-            Optional<Task> locked = nextTasksToLock.stream()
-                    .map(taskRepository::lock)
-                    .filter(Objects::nonNull)
-                    .findFirst();
-
-            if (locked.isPresent()) {
-                return locked.get();
+            if (nextTasksToLock.isEmpty()) {
+                return Collections.emptyList();
             }
 
             if (Thread.currentThread().isInterrupted()) {
-                break;
+                return Collections.emptyList();
             }
-        } while (!(nextTasksToLock = findNextForExecution(processableTasks, limit)).isEmpty());
 
-        return null;
+            do {
+                List<Task> locked = taskRepository.lock(con, nextTasksToLock);
+
+                if (locked.size() > 0) {
+                    return locked;
+                }
+
+                if (Thread.currentThread().isInterrupted()) {
+                    break;
+                }
+            } while (!(nextTasksToLock = findNextForExecution(con, processableTasks, limit)).isEmpty());
```

Comment:
This loop is no longer required, because it is checked outside of this function on calling site.